src/lib.rs

```diff
@@ -1,130 +1,206 @@
-use lazy_static::lazy_static;
-
 extern crate wapc_guest as guest;
 use guest::prelude::*;
 
 use k8s_openapi::api::core::v1 as apicore;
 
 extern crate kubewarden_policy_sdk as kubewarden;
-use kubewarden::{logging, protocol_version_guest, request::ValidationRequest, validate_settings};
+use kubewarden::{protocol_version_guest, request::ValidationRequest, validate_settings};
 
 mod settings;
 use settings::Settings;
 
-use slog::{info, o, warn, Logger};
-
-lazy_static! {
-    static ref LOG_DRAIN: Logger = Logger::root(
-        logging::KubewardenDrain::new(),
-        o!("policy" => "sample-policy")
-    );
-}
-
 #[no_mangle]
 pub extern "C" fn wapc_init() {
     register_function("validate", validate);
     register_function("validate_settings", validate_settings::<Settings>);
     register_function("protocol_version", protocol_version_guest);
 }
 
+#[derive(Debug, PartialEq)]
+enum PolicyResponse {
+    Accept,
+    Reject(String),
+}
+
 fn validate(payload: &[u8]) -> CallResult {
     let validation_request: ValidationRequest<Settings> = ValidationRequest::new(payload)?;
 
-    info!(LOG_DRAIN, "starting validation");
-
-    // TODO: you can unmarshal any Kubernetes API type you are interested in
-    match serde_json::from_value::<apicore::Pod>(validation_request.request.object) {
-        Ok(pod) => {
-            // TODO: your logic goes here
-            if pod.metadata.name == Some("invalid-pod-name".to_string()) {
-                let pod_name = pod.metadata.name.unwrap();
-                info!(
-                    LOG_DRAIN,
-                    "rejecting pod";
-                    "pod_name" => &pod_name
-                );
-                kubewarden::reject_request(
-                    Some(format!("pod name {} is not accepted", &pod_name)),
-                    None,
-                )
-            } else {
-                info!(LOG_DRAIN, "accepting resource");
-                kubewarden::accept_request()
-            }
-        }
-        Err(_) => {
-            // TODO: handle as you wish
-            // We were forwarded a request we cannot unmarshal or
-            // understand, just accept it
-            warn!(LOG_DRAIN, "cannot unmarshal resource: this policy does not know how to evaluate this resource; accept it");
-            kubewarden::accept_request()
-        }
+    let pod = match serde_json::from_value::<apicore::Pod>(validation_request.request.object) {
+        Ok(pod) => pod,
+        Err(_) => return kubewarden::accept_request(),
+    };
+
+    match do_validate(&pod) {
+        PolicyResponse::Accept => kubewarden::accept_request(),
+        PolicyResponse::Reject(msg) => kubewarden::reject_request(Some(msg), None),
+    }
+}
+
+fn do_validate(pod: &apicore::Pod) -> PolicyResponse {
+    if pod.spec.is_none() {
+        return PolicyResponse::Accept;
+    }
+
+    let pod_spec = pod.spec.clone().unwrap();
```

Comment:
In all policies targeting pods (specially those meant to replace PSP's) we should behave in the same way when it comes to a missing `.pod.spec`. Accepting as this policy does is one option. Converting the option to a result and bubbling the error case is another option as done in https://github.com/kubewarden/selinux-psp-policy/blob/2a32fc9e6ee7e3460222374591ac8d6f4aea6fec/src/lib.rs#L46.

Honestly, I don't have a strong opinion, since I think having a pod without a spec (in that it's not present, not even `spec: {}`) is a corner case. I don't know what's the usage of that TBH.

In any case, I think it would be beneficial that all our policies for pods behave in the same way in this case.

Reply:
In this case, not having a pod spec means there's no container that has a writable root filesytem :)

But I get what you mean